### Regra de Simpson simples

In [1]:
import numpy as np
#Número de sub-intervalos [MÉTODO SIMPLES N=2 !]
N=2

# ============================ INTERVALO ===================#
#Extrema esquerda do intervalo
a=0.0
#Extrema direita do intervalo
b=2.0
#Comprimento de cada intervalo
h=(b-a)/N

#Função cuja integral será estimada numericamente
# ======================== FUNÇÃO ===================== #
def f(x):
    y=x**4
    return y

X=np.zeros(N+1)
# Preenchendo o vetor X com as abcissas dos nodos de interpolação .
# O loop é exclusivo à direita
for i in np.arange(0,N+1,1):
    X[i]=a+i*h
# Gerando a ordenada de cada um desses nodos de interpolação.   
Y=f(X)

#MATRIZ DE vandermonde
A= np.zeros([(N+1),(N+1)])
# UMA MATRIZ N+1 x N+1 foi criada

# Inserindo no vetor X |INSERINDO NA FILA 0 | O ELEMENTO [1] | ao longo de AXIS 0  -> FILA=LINHA
#TODOS OS OUTROS ELEMENTOS SÃO SHIFTADOS ADIANTE [NO SENTIDO CRESCENTE DO ÍNDICE].
L=np.insert(X,0,[1],axis=0)
del(L)
#foi desnecessário , mas vale a observação.

#copiando todo o vetor na coluna da matriz A
for i in np.arange(0,N+1,1):
    A[:][i]=np.copy(X**i)
    
#PORTANTO O VETOR Y foi construído na forma de vetor-coluna !

v= np.zeros(N+1)
for i in np.arange(0,N+1,1):
    k=i+1
    v[i]= (   ( (b**k)-(a**k) )/ ( k )   )
    
# Resolvendo a EQUAÇÃO LINEAR Ax = v
# no caso r = x
r = np.linalg.solve(A, v)

I_s=np.dot(np.transpose(r),Y)
print("O resultado pela Regra de Simpson simples foi I_s = %.6f" %(I_s))

O resultado pela Regra de Simpson simples foi I_s = 6.518519


Logo, o resultado pela Regra de Simpson zimples foi $I_{s}(x^{4}) = 6.518519$ no intervalo $[0,2]$.

In [2]:
#DELETANDO FORÇADAMENTE
%reset -f

### Regra de Simpson composta

In [3]:
import numpy as np
# intervalo [a,b]
# ============================ INTERVALO ===================#
#Extrema esquerda do intervalo
a=0
#Extrema direita do intervalo
b=2
#Número de sub-intervalos.
#------------> PARA A Regra de Simpson composta DEVE SER PAR <-------------------------#
N=20

#Diferença entre as abcissas de dois nós adjacentes.
h=(b-a)/N
#Vetor cujas entradas são os abcissas dos nós.
X=np.zeros(N+1)

#Loops são não inclusivos à direita.
for i in np.arange(0,N+1,1):
    X[i]= (a + i*h)
# ======================== FUNÇÃO ===================== #
def f(x):
    y= x**4
    return y

Y=f(X)

k=N/2

impar=0.0
par=0.0
# contabilizando o somatorio de índices par
# loop é não inclusivo à EXTREMA DIREITA
for i in np.arange(2,N,2): 
    par = Y[i] + par
# contabilizando o somatorio de índices ímpar
# loop é não inclusivo à EXTREMA DIREITA
for i in np.arange(1,N-1,2): 
    impar=Y[i] + impar
#Integração por Regra de Simpson composta
S_N= (h/3)*( Y[0] + Y[N] +4*impar +2*par  )    

print("O resultado pela Regra de Simpson composta foi:\nS_20 = %.6f" %(S_N))

O resultado pela Regra de Simpson composta foi:
S_20 = 4.662413


Logo, o resultado pela Regra de Simpson composta foi $I_{20}(x^{4}) = 4.662413$ no intervalo $ [0,2]$&nbsp;.

In [2]:
#DELETANDO FORÇADAMENTE
%reset -f